***
# **Customer Purchase Behavior Analysis and Reporting**
***

## **Data Extraction**

In [3]:
# Importing sqlalchemy
import sqlalchemy

In [ ]:
# Installation of pymysql
!pip install pymysql

In [5]:
import pymysql

In [6]:
import pandas as pd

In [7]:
# Creating connection between SQL and Python
engine = sqlalchemy.create_engine('mysql+pymysql://root:2001@localhost:3306/construct_week_project')

In [8]:
#Loading customers, products, purchase tables from SQLDB
customers = pd.read_sql_table("customers",engine)
products = pd.read_sql_table("products",engine)
purchase = pd.read_sql_table("purchase",engine)


## **Data Analysis**

#### *Statistics*

In [11]:
# Calculating total purchase quantity
Total_Purchase_Quantity = sum(purchase["PurchaseQuantity"])
Total_Purchase_Quantity

3053

In [12]:
# Calculating total revenue
Total_Revenue = sum(purchase["PurchasePrice"])
Total_Revenue

489274.72

In [13]:
# Calculating average purchase value
Average_Purchase_Value = purchase["PurchasePrice"].mean()
Average_Purchase_Value

489.27471999999995

#### *Identify top customers and their purchasing behavior*

In [15]:
merge1 = pd.merge(customers, purchase, on = "CustomerID")
merge2 = pd.merge(merge1, products, on = "ProductID")
top_customers = merge2.groupby(["CustomerID", "CustomerName","PurchaseDate",
                                "ProductName", "PurchaseQuantity","ProductCategory"]).sum()["PurchasePrice"].sort_values(ascending=False)
top_customers.head(10)

CustomerID  CustomerName      PurchaseDate  ProductName     PurchaseQuantity  ProductCategory
778         Jillian Baker     2024-03-21    Vacuum Cleaner  2                 Home Appliances    999.98
268         Patricia Taylor   2023-09-26    Smartphone      5                 Electronics        999.44
282         Mikayla Rios      2023-10-04    Camera          5                 Electronics        998.17
74          Matthew Espinoza  2023-07-24    Television      3                 Electronics        997.45
504         Jamie Montoya     2023-12-21    Headphones      4                 Electronics        996.01
872         Charles Clark     2024-05-03    Smartphone      3                 Electronics        995.97
460         Antonio Hughes    2023-12-05    Vacuum Cleaner  4                 Home Appliances    994.58
927         Carlos Smith      2024-05-24    Headphones      5                 Electronics        993.55
816         Steven Rivera     2024-04-08    Refrigerator    2             

#### *Analyze purchase trends over time (monthly, quarterly, yearly)*

In [17]:
purchase['Month'] = purchase["PurchaseDate"].dt.month
purchase['Year'] = purchase["PurchaseDate"].dt.year
purchase['Quarter'] = purchase["PurchaseDate"].dt.quarter
purchase

TransactionID  CustomerID  ProductID PurchaseDate  PurchaseQuantity  \
0                1         726          1   2024-03-01                 5   
1                2         620          2   2024-01-26                 4   
2                3         904         11   2024-05-13                 3   
3                4         249          3   2023-09-21                 5   
4                5         617         12   2024-01-25                 4   
..             ...         ...        ...          ...               ...   
995            996         552         11   2024-01-03                 2   
996            997         385         10   2023-11-11                 5   
997            998         364         15   2023-11-04                 5   
998            999         366          5   2023-11-05                 4   
999           1000         925          4   2024-05-22                 5   

     PurchasePrice  Month  Year  Quarter  
0           689.99      3  2024        1  
1            79.27      1  2024        1  
2           666.75      5  2024        2  
3           316.19      9  2023        3  
4           700.24      1  2024        1  
..             ...    ...   ...      ...  
995         796.67      1  2024        1  
996         271.88     11  2023        4  
997         594.69     11  2023        4  
998         140.89     11  2023        4  
999         861.00      5  2024        2  

[1000 rows x 9 columns]

##### 1. Monthly

In [19]:
avg_monthly_sales = purchase.groupby("Month")["PurchasePrice"].mean()
avg_monthly_sales

Month
1     449.228247
2     486.985529
3     542.647246
4     517.661212
5     509.523452
6     471.103134
7     515.663333
8     434.980000
9     502.847312
10    498.298312
11    419.360947
12    535.158969
Name: PurchasePrice, dtype: float64

##### 2. Quarterly

In [21]:
avg_quarterly_sales = purchase.groupby("Quarter")["PurchasePrice"].mean()
avg_quarterly_sales

Quarter
1    487.695498
2    500.136037
3    487.509316
4    483.712565
Name: PurchasePrice, dtype: float64

##### 3. Yearly

In [23]:
avg_yearly_sales = purchase.groupby("Year")["PurchasePrice"].mean()
avg_yearly_sales

Year
2023    483.141934
2024    496.561597
Name: PurchasePrice, dtype: float64

#### *Identify the top-performing product categories*

In [25]:
merge3 = pd.merge(purchase, products, on = "ProductID")
Category_Sales = merge3.groupby("ProductCategory")["PurchasePrice"].sum().sort_values(ascending=False)
Category_Sales

ProductCategory
Electronics        248194.56
Home Appliances    241080.16
Name: PurchasePrice, dtype: float64

#### *Identify the top-performing products*

In [27]:
Product_Sales = merge3.groupby("ProductName")["PurchasePrice"].sum().sort_values(ascending=False)
Product_Sales.head(10)

ProductName
Microwave          30970.75
Smartwatch         29407.68
Toaster            29116.03
Heater             28828.67
Headphones         27748.78
Refrigerator       27730.95
Printer            27360.29
Television         25188.60
Camera             24985.70
Air Conditioner    24812.05
Name: PurchasePrice, dtype: float64

***
## **Summary Report**
***

### Assumptions ###
While analysing the data I have assumed that all the customers who made purchases are unique and have made purchase only once. So, accordingly I have
given unique IDs to all the customers in the dataset.

 *Reason: 
According to the data provided, even customers with the same name are from different countries. Because of that, I have assumed that they are from different customers with same name.*

### Statistics

- ##### Total Purchased Quantity:
  3053

- ##### Total Revenue:
  489274.72

- ##### Average Purchase Value:
  489.27


### Top customers and their Purchase Behavior ###

- ##### Top Customers:
   Jillian Baker, Patricia Taylor, Mikayla Rios, Matthew Espinoza, Jamie Montoya, Charles Clark, Antonio Hughes, Carlos Smith,
                  Steven Rivera, Casey Johnson

- ##### Purchase Behavior:
  -  Electronics Category Dominance:  Most top customers purchased electronics, indicating a preference for high-value and tech-related products
                                    like smartphones, cameras, and headphones.
                                    
  -  High Purchase Frequency:  Customers like Patricia Taylor, Mikayla Rios, and Carlos Smith make multiple purchases over time, 
                                    suggesting they are regular buyers.
                                           
  -  Diverse Consumer Behavior:  They vary in terms of purchase frequency and product preferences, reflecting diverse consumer behaviors 
                                    within the electronics and home appliances categories.

    
### Purchase Trends over Time ###

- ##### Monthly:
  The analysis suggests that consumer spending does not have much variations throughout the year, with a mild increase in March, April and Decemeber.

- ##### Quarterly:
  The consumer spending has been constant throughout the year, with a little increase in the second quarter. 
  Last quarter includes a slight dip  in consumer spending when compared to others.

- ##### Yearly:
  The yearly sales has a slight increase in the year 2024 based on the current given data.


### Top-Performing Product Categories ###

- ##### Electronics: 
  With a total sales amount of 2,48,194.56, the Electronics category emerges as the top-performing category based on sales. This category includes 
  products like smartphones, cameras, televisions, and headphones, which are typically higher in value and often have a higher profit margin.


### Top-Performing Products ###

  Consumers are spending more on luxury products like Microwave, Smartwatch and Toaster than some other necessary products like smartphones 
  and washing machine.



### **Conclusion and Key Insights** ###

- Electronics Category Dominance
- Diverse Consumer Behavior
- Luxury Products

Understanding these patterns can help in tailoring marketing strategies and product offerings to better serve the customer segments.
Businesses can use these insights to focus on optimizing their offerings, marketing strategies, and inventory management within the Electronics 
category and luxury products to capitalize on consumer preferences and maximize profitability throughout the year.